### ISSUE TITLE
OSDI 26 | Teaching The Old Dog New Tricks: Building Efficient Data Pipelines for Large-Scale LLM Pre-training (Operational Systems)

### ISSUE THREAD

**Issue by skjsjhb** [AI-Infra-Team/awesome-papers#215]
Teaching The Old Dog New Tricks: Building Efficient Data Pipelines for Large-Scale LLM Pre-training (Operational Systems)

*Luofan Chen and Chenhan Wang, University of Science and Technology of China and ByteDance Seed; Weidong Zhang, Jinxin Chi, Hequan Zhang, Zanbo Wang, Chenyuan Wang, Lishu Luo, Sijin Wu, Junqi Hu, Jun Wang, and Cheng Chen, ByteDance Seed; Lixin Huang, Liyang Zhao, Yong Tian, and Jun Guo, ByteDance; Youhui Bai, University of Science and Technology of China; Wencong Xiao, ByteDance Seed; Kang Chen, Tsinghua University; Cheng Li, University of Science and Technology of China and Institute of Artificial Intelligence, Hefei Comprehensive National Science Center*

Paper PDF: <https://www.usenix.org/system/files/osdi26-chen-luofan.pdf>

## Motivation

Data pipeline is less optimized than computing in training.

3 major problems:

- **Cross-DC Latency Trap**: Evaluation runs on remote, WAN transporting lags, wastes GPU.
- **Initialization I/O Storm**: Training startup/restart reads hot files, creates tail latencies.
- **Transformation Wall**: CPU transformation of data does not keep up with training.

> We argue that these bottlenecks persist because storage systems operate **reactively** while training workloads are inherently **deterministic**.

Optimizations: Predictive Checkpoint Replication, Proactive Hotspot Prediction, Storage-Side Transformation Offloading

Works even for legacy big data systems e.g. HDFS.

### Cross-DC Companion Evaluation

<img width="418" height="250" alt="Image" src="https://github.com/user-attachments/assets/98ec0287-12bf-4c96-a949-2049143c650a" />

Companion Evaluation Workflow:

1. Merge distributed ckpt shards for inference.
2. Reshard model for the evaluation device.
3. Standard model serving + evaluation.

Bottleneck:

~55.6% of evaluation time sped on data transporting, among them ~84.8% comes from merging shards, i.e. read tensors remotely. This lags because:

- Small tensor I/O is inefficient: 
  - Read amplification: Even reading a 4KB tensor would require a larger block in HDFS.
  - RTT-bound throughput: Latency affects small tensor reads severely.
- Bandwidth contention:
  - Current policy does not distinguish between critical path and background tasks.

Measures:

- Predictive checkpoint replication:
  1. Evaluation activity is predictable.
  2. Initiate a pipelined packing of shards right before evaluation, once saved.
  3. Forward the contiguous stream to evaluation.
- Signal-driven prioritization:
  - Sends priority signal to sort urgent tasks, e.g. periodical evaluation vs. training anomalies.

Evaluation: -76.1% latency of merging shards, effectively 2M GPU hrs

### Initialization I/O Storm

Startup/restart of training includes sync loading. Blocking I/O operations during file download are
the primary bottleneck (~39.95%).

Bottleneck:
- Analysis reveals that data contention is the primary culprit, rather than metadata.
- Hot files: top 5% accounts for 38.8% total.

The problem is due to a tradeoff: parallel saving + redundancy elimination for writing performance, but this harms reading as the traffic involves one-to-many transports.

Measures:

- Traditional solution: detect hot files and replicate. This is flawed:
  - By the time the file is cached, the loading is over.
  - Initiating caching further jams the traffic.
- Let training framework explicitly informs the storage system about upcoming access patterns.
  - The set of hot files is known.
- Classifies data as 2: global metadata + replicated metadata
  - The storage system define appropriate replicas for them.

Evaluation: Eliminating straggler reads = 40.8% improvement

### MM-L Transformation Wall

Data loading can be hidden behind GPU for text models, but not for multimodal ones.

Workflow of a dataloader:

1. Metadata resolution.
2. Data retrieval.
3. Sample transformation. Runs on CPU to avoid affecting training.

Transformation is critical, contributes 94.4% in total.

Bottleneck:

- The heterogeneous nature of multimodal data. Transformation cost varies among texts and videos.
- An MM-L batch contains short/long videos, workload varies. Difficult to hide.

Measures:

- Offline transformation is appealing, but causes extreme storage amplification (40x expansion), and does not fit into dynamic requirements.
- Exploit: Storage nodes in high-throughput clusters often have underutilized CPU capacity. Use them.
  - Storage nodes are bounded by HDD seeks, NIC limits, etc. not computing.
- Schedule synchronization:
  1. Training framework knows what data might come next.
  2. Training machine tells storage nodes to resolve upcoming data blocks.
  3. Prepared data is sent back.
- Just-in-Time (JIT) transformation: speculatively read raw binary data and execute the registered transformation graph. Sampling is also done, or the network jams.
- Load-aware fallback: switches to read-only mode if a node is overloaded.

Evaluation: -85.7% P99 latency, -94% CPU usage on training hosts
